odxtools/progcode.py

```diff
@@ -52,4 +56,9 @@ def _resolve_odxlinks(self, odxlinks: OdxLinkDatabase) -> None:
         pass
 
     def _resolve_snrefs(self, diag_layer: "DiagLayer") -> None:
-        pass
+        db = diag_layer.database
```

Comment:
It would be better to add db as argument to `_resolve_snrefs`, rather than a property of the DiagLayer

Reply:
Most SNREFs can only be resolved if the applicable diagnostic layer is known (e.g. TABLE-SNREFs, DIAG-COMM-SNREFs, etc.), but it is impossible to determine the relevant diaglayer given only the database: e.g., when resolving SNREFs within a service, it is completely unclear which of the diag layers that the respective service is part of ought to be used.

We probably should to rethink the SNREF resolution mechanism anyway: SNREFs lead to a really bad situation with inheritance where the object that is referenced depends on how a given object is accessed (cf #200). Example:

- DiagLayer `B` ("Base") defines a request `R` which references a DOP `Foo` via a short name reference
- Two DiagLayers `D1` and `D2` are derived from `B`. Both `D1` and `D2` redefine the DOP `Foo`
- In this case, `D1.R.Foo` ought to be different from `D2.R.Foo` even though `R` is the same object in both cases.

So far, I have not come up with an idea of how to implement this in a good way. The only solution which I've come up with is to make all objects in a database use a given diaglayer for their SNREFs, but that invalidates the database for any other diaglayer...

Reply:
I am talking about adding it as an extra argument
```py
def _resolve_snrefs(self, diag_layer: "DiagLayer", database: "Database") -> None:
```
A more generic approach would be
```py
def _resolve_snrefs(self, lookup) -> None:
```
`lookup` will provide two functions `resolve_snref(name)` and `get_file(name)`

the implementation of those two functions could be adapted later to account for corner cases like #200

Reply:
I basically agree. I would propose to implement this by merging `_resolve_snrefs()` and `_resolve_odxlinks()` into `_resolve_references()` and passing it a context object which provides access to all relevant objects (OdxLinkDatabase, Database, Diaglayer, etc). This is quite a bit of work, though, and IMO thus a bit out of scope of this PR (i.e., improving the auxiliary file handling). An issue I can see with this approach is that the state these context objects must exhibit in order to call these methods is quite magical...

Reply:
another issue might be properly dealing with inheritance.